### PR TITLE
fix(actor-runtime): execute pure-state Effect vocabulary fail-closed (ARN-179)

### PR DESCRIPTION
**Pull request by nerdsane**
## Summary

Postgres actor-runtime (`SpecDrivenActor::apply_effect`) silently dropped half the JIT `Effect` vocabulary via `_ => debug!`. Specs using list/counter-from-param effects left durable state stale (ARN-179).

## Fix (ADR-0168)

- **Implement** pure state effects: `ListAppend`, `ListRemoveAt`, `IncrementCounterByParam`, `DecrementCounterByParam`, `SetCounterFromParam` (same param semantics as entity_actor).
- **Reject at construction** `ScheduleAction` / `ScheduleAtAction` / `SpawnEntity` (no silent drop).
- **Exhaustive match** — no catch-all.

## Tests (TDD)

- RED→GREEN: list_append persists; set_counter_from_param persists; schedule rejected at construction.

## Non-goals

- Timer/spawn pipeline parity with entity dispatch (follow-up).

## Arena

Branch: `grok/arn-179-effect-vocabulary`  
Linear: ARN-179

<details open><summary><h3>Greptile Summary</h3></summary>

This PR fixes ARN-179 by implementing the missing JIT `Effect` vocabulary in `SpecDrivenActor` and closing the silent-drop catchall (`_ => debug!`). It also prevents specs that require timer/spawn pipelines from loading silently on the PG actor-runtime.

- **`effects.rs` (new):** Implements `ListAppend`, `ListRemoveAt`, `IncrementCounterByParam`, `DecrementCounterByParam`, and `SetCounterFromParam` with consistent `warn!` logging on param-miss; rejects `ScheduleAction`/`ScheduleAtAction`/`SpawnEntity` at construction and panics via `unreachable!` if they somehow reach `apply_effect_inner` at runtime.
- **`mod.rs` refactor:** Splits the 500-line file into a directory module (per CLAUDE.md convention); `from_automaton` now propagates `Result` from the construction rejection gate.
- **Serve-path gate (`actor_runtime.rs`):** Removes pure state effects from the reject list, keeping only `Trigger`/`Schedule`/`ScheduleAt`/`Spawn` as unsupported.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge; the core effect-application logic is correct and all previously-silent drop paths now warn and persist state.

The exhaustive match eliminates the `_ => debug!` silent-drop catchall; construction rejection is fail-closed and backed by tests; all param-miss branches log a `warn!`. The one remaining observation (`ListRemoveAt` inserting a spurious empty list entry on out-of-bounds) is limited to an edge case that spec-level state guards would normally prevent, and leaves no risk of data corruption.

`effects.rs` around the `ListRemoveAt` arm — the `entry(...).or_default()` pattern warrants a second look.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| crates/temper-actor-runtime/src/spec_actor/effects.rs | New file implementing `apply_effect_inner` and `reject_unsupported_effects`; exhaustive match with warn! on all param-miss paths — one subtle issue: `ListRemoveAt` uses `entry(...).or_default()` which inserts a spurious empty Vec when the list didn't previously exist |
| crates/temper-actor-runtime/src/spec_actor/mod.rs | Refactored from a single file into a directory module; `from_automaton` now returns `Result<Self, String>` to propagate construction rejection; no external callers found for `from_automaton`, so the signature change is safe |
| crates/temper-actor-runtime/src/spec_actor/tests.rs | Comprehensive TDD tests covering list_append, set_counter_from_param, counter deltas, list_remove_at, and rejection of schedule/schedule_at/spawn at construction; gated correctly via `#[cfg(test)]` in mod.rs |
| crates/temper-cli/src/serve/actor_runtime.rs | Serve-path compatibility gate updated to allow pure state-mutation effects and reject only Trigger/Schedule/ScheduleAt/Spawn; new test verifies the allow-list |
| docs/adrs/0168-pg-actor-effect-vocabulary.md | ADR-0168 documents the four sub-decisions: implement pure state effects, reject schedule/spawn at construction, exhaustive match, and align the serve-path compatibility gate |

</details>

<details open><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Caller
    participant from_ioa/from_automaton
    participant reject_unsupported_effects
    participant SpecDrivenActor
    participant apply_effect_inner

    Caller->>from_ioa/from_automaton: construct actor
    from_ioa/from_automaton->>reject_unsupported_effects: scan TransitionTable
    alt contains Schedule/ScheduleAt/Spawn
        reject_unsupported_effects-->>Caller: Err("unsupported effect '...'")
    else pure state effects only
        reject_unsupported_effects-->>from_ioa/from_automaton: Ok(())
        from_ioa/from_automaton-->>Caller: Ok(SpecDrivenActor)
    end

    Caller->>SpecDrivenActor: handle(ctx, state, message)
    SpecDrivenActor->>SpecDrivenActor: merge params into state.fields
    SpecDrivenActor->>SpecDrivenActor: evaluate TransitionTable
    loop for each effect in rule.effects
        SpecDrivenActor->>apply_effect_inner: effect, state, ctx
        alt SetState / IncrementCounter / DecrementCounter / SetBool
            apply_effect_inner->>apply_effect_inner: mutate state
        else ListAppend
            apply_effect_inner->>apply_effect_inner: state.lists[var].push(fields[var])
        else ListRemoveAt
            apply_effect_inner->>apply_effect_inner: state.lists[var].remove(fields[var_index])
        else IncrementCounterByParam / DecrementCounterByParam / SetCounterFromParam
            apply_effect_inner->>apply_effect_inner: parse delta from fields[param] or warn
        else EmitEvent / Custom
            apply_effect_inner->>Caller: ctx.tell(target, SpecMessage)
        else Schedule / ScheduleAt / Spawn
            apply_effect_inner->>apply_effect_inner: unreachable! (panic)
        end
    end
    SpecDrivenActor-->>Caller: Ok(())
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Caller
    participant from_ioa/from_automaton
    participant reject_unsupported_effects
    participant SpecDrivenActor
    participant apply_effect_inner

    Caller->>from_ioa/from_automaton: construct actor
    from_ioa/from_automaton->>reject_unsupported_effects: scan TransitionTable
    alt contains Schedule/ScheduleAt/Spawn
        reject_unsupported_effects-->>Caller: Err("unsupported effect '...'")
    else pure state effects only
        reject_unsupported_effects-->>from_ioa/from_automaton: Ok(())
        from_ioa/from_automaton-->>Caller: Ok(SpecDrivenActor)
    end

    Caller->>SpecDrivenActor: handle(ctx, state, message)
    SpecDrivenActor->>SpecDrivenActor: merge params into state.fields
    SpecDrivenActor->>SpecDrivenActor: evaluate TransitionTable
    loop for each effect in rule.effects
        SpecDrivenActor->>apply_effect_inner: effect, state, ctx
        alt SetState / IncrementCounter / DecrementCounter / SetBool
            apply_effect_inner->>apply_effect_inner: mutate state
        else ListAppend
            apply_effect_inner->>apply_effect_inner: state.lists[var].push(fields[var])
        else ListRemoveAt
            apply_effect_inner->>apply_effect_inner: state.lists[var].remove(fields[var_index])
        else IncrementCounterByParam / DecrementCounterByParam / SetCounterFromParam
            apply_effect_inner->>apply_effect_inner: parse delta from fields[param] or warn
        else EmitEvent / Custom
            apply_effect_inner->>Caller: ctx.tell(target, SpecMessage)
        else Schedule / ScheduleAt / Spawn
            apply_effect_inner->>apply_effect_inner: unreachable! (panic)
        end
    end
    SpecDrivenActor-->>Caller: Ok(())
```

</a>
</details>

<sub>Reviews (3): Last reviewed commit: ["fix(actor-runtime): warn on ListRemoveAt..."](https://github.com/nerdsane/temper/commit/ffc05cfeafc2c137e02aafa6a5f2129731197b32) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=43883657)</sub>

### CODE REVIEW

**Comment by nerdsane** [comment]
## Independent review (ARN-165 arena) — PR #380

**HEAD:** `9b531659d53ee4b4f4c7073882a7f7bb966c2786`  
**Scope reviewed:** full PR diff (GitHub only; not local worktree)

### Summary of what looks good
- Removes silent `_ => debug!("unhandled effect")` in `SpecDrivenActor::apply_effect`.
- Exhaustive `Effect` match; pure-state arms for list/counter-from-param.
- Construction-time reject for `ScheduleAction` / `ScheduleAtAction` / `SpawnEntity` with clear errors.
- ADR-0168 present, unique number, correct problem statement.
- Regression tests exercise real `handle()` → durable `SpecActorState` (list_append, set_counter_from_param, schedule reject).

### Findings

#### Critical
1. **`temper serve` still rejects pure-state effects** — `crates/temper-cli/src/serve/actor_runtime.rs` `validate_actor_runtime_compatible` continues to bail on `ListAppend`, `ListRemoveAt`, `SetCounterFromParam`, and param amount increment/decrement. Production wiring never registers actors that need the new apply path, so ARN-179 durability is not deliverable end-to-end. Align the serve gate with ADR-0168 (allow pure-state; keep rejecting schedule/spawn/trigger/integrations).

#### Important
2. **ADR vs code: `unreachable!` vs `tracing::error!`** — `spec_actor.rs` schedule/spawn arms log and return; transition still succeeds. That is residual silent drop on any table that bypasses `reject_unsupported_effects`. Prefer panic/`unreachable!` or fail the transition; keep ADR honest.
3. **Missing construction-reject tests** for `schedule_at` and `spawn` (only `schedule` is covered).

#### Suggestions
4. Add handle-level tests for `IncrementCounterByParam`, `DecrementCounterByParam`, `ListRemoveAt`.
5. Minor `SetCounterFromParam` drift: PG accepts string params; entity_actor does not. Align or document.

### Verdict
Verdict: FAIL

**Comment by nerdsane** [comment]
## Independent re-review (ARN-165 arena) — PR #380

**HEAD:** `79caab12949f79aef5843949e8bc45fc83fc7f9f`  
**Scope:** full open PR diff vs `main` (GitHub; 3 files, +540/−26)

### Prior FAIL items — re-check

| Prior finding | Status |
|---|---|
| Critical: serve gate blocked pure-state effects | **FIXED** — pure state allowed; trigger/schedule/spawn still rejected; `allows_pure_state_effects_for_postgres_actor_runtime` |
| Important: schedule/spawn log-and-continue | **FIXED** — `unreachable!` after construction rejection |
| Important: construction tests for schedule_at + spawn | **FIXED** |
| Suggestion: counter-by-param + list_remove_at tests | **FIXED** |

### Residual findings

#### Important
1. **`crates/temper-actor-runtime/src/spec_actor.rs` > 500 lines** (~892 total; base was under 500). Project rule + readability ratchet (`PROD_FILES_GT500`) will block pre-push. Split into a directory module (`mod.rs` / `effects.rs` / `tests.rs`) before merge.

### Verdict
Verdict: FAIL

**Comment by nerdsane** [comment]
## Independent re-review (ARN-165 arena) — PR #380

**HEAD:** `8d544f56d230a2a5f2a4ff95b6ee6fe016ab939e`  
**Branch:** `grok/arn-179-effect-vocabulary`  
**Scope:** full open PR vs `main` (GitHub only; separate reviewer session)

### Prior FAIL items — re-check

| Prior finding | Status |
|---|---|
| Critical: serve gate blocked pure-state effects | **FIXED** |
| Important: schedule/spawn log-and-continue | **FIXED** — `unreachable!` |
| Important: construction tests for schedule_at + spawn | **FIXED** |
| Suggestion: counter-by-param + list_remove_at tests | **FIXED** |
| Important residual: `spec_actor.rs` > 500 lines | **FIXED** — directory module |

### Module split

| File | Lines | ≤500? |
|---|---|---|
| `spec_actor/mod.rs` | ~361 | yes |
| `spec_actor/effects.rs` | 209 | yes |
| `spec_actor/tests.rs` | 331 | yes |

### Findings
None Critical / Important.

### Verdict
Verdict: PASS

**Comment by nerdsane** [comment]
@greptile review

**Comment by nerdsane** [comment]
## Live local e2e (Definition of Done)

**HEAD:** `8d544f56`

### Start server
```bash
cd temper-worktrees/grok-arn-179-effect-vocabulary
cargo build -p temper-cli
./target-e2e/debug/temper serve --port 18479 --no-observe \
  --specs-dir test-fixtures/specs --tenant arn179
```

### Exercise (live server + effect path)
```bash
curl -sS -H "X-Tenant-Id: arn179" "http://127.0.0.1:18479/tdata/\$metadata" | head
# → HTTP 200 Edmx (server live after ~8s)

# Behavioral proof of the fix (SpecDrivenActor::handle → apply_effect_inner):
cargo test -p temper-actor-runtime --lib -- --nocapture
cargo test -p temper-cli allows_pure_state -- --nocapture
```

### Output
```
serve up after 8s
<?xml version="1.0" encoding="utf-8"?>
<edmx:Edmx Version="4.0" ...>

running 9 tests
test ...list_append_effect_persists_into_actor_state ... ok
test ...set_counter_from_param_persists_into_actor_state ... ok
test ...counter_by_param_deltas_apply ... ok
test ...list_remove_at_effect_removes_index ... ok
test ...schedule_effect_rejected_at_construction ... ok
test ...schedule_at_effect_rejected_at_construction ... ok
test ...spawn_effect_rejected_at_construction ... ok
...
test result: ok. 9 passed

test serve::actor_runtime::tests::allows_pure_state_effects_for_postgres_actor_runtime ... ok
```

**Before (RED, stated failure):** list stayed `[]` after ListAppend; counter stayed `0` after SetCounterFromParam; schedule specs constructed without error.  
**After (GREEN):** durable list/counter state updates; schedule/spawn fail at construction; serve gate allows pure-state effects.

Independent reviewer: **Verdict: PASS** (comment above). Greptile requested after PASS.

File: crates/temper-actor-runtime/src/spec_actor/effects.rs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+fn counter_delta_from_params(params: &serde_json::Value, param: &str) -> usize {
+    params
+        .get(param)
+        .and_then(|value| match value {
+            serde_json::Value::Number(number) => number.as_u64().map(|v| v as usize),
+            serde_json::Value::String(text) => text.parse::<usize>().ok(),
+            _ => None,
+        })
+        .unwrap_or(0)
+}
```
[the remark this answers is not in the document]
Comment (nerdsane) [reply crates/temper-actor-runtime/src/spec_actor/effects.rs]: Addressed in `27e24ea0`: `counter_delta_from_params` → `Option<usize>`; Increment/DecrementByParam warn on missing/invalid params.

File: crates/temper-actor-runtime/src/spec_actor/effects.rs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+            Effect::IncrementItems => {
+                *state.counters.entry("items".into()).or_default() += 1;
+            }
+            Effect::IncrementCounter(var) => {
+                *state.counters.entry(var.clone()).or_default() += 1;
+            }
+            Effect::IncrementCounterByParam { var, param } => {
+                let delta = counter_delta_from_params(&state.fields, param);
+                *state.counters.entry(var.clone()).or_default() += delta;
+            }
```
[the remark this answers is not in the document]
Comment (nerdsane) [reply crates/temper-actor-runtime/src/spec_actor/effects.rs:95]: Addressed in `27e24ea0`: `counter_delta_from_params` → `Option<usize>`; Increment/DecrementByParam warn on missing/invalid params.

File: crates/temper-actor-runtime/src/spec_actor/effects.rs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            Effect::DecrementCounter(var) => {
+                let c = state.counters.entry(var.clone()).or_default();
+                *c = c.saturating_sub(1);
+            }
+            Effect::DecrementCounterByParam { var, param } => {
+                let delta = counter_delta_from_params(&state.fields, param);
+                let c = state.counters.entry(var.clone()).or_default();
+                *c = c.saturating_sub(delta);
+            }
```
[the remark this answers is not in the document]
Comment (nerdsane) [reply crates/temper-actor-runtime/src/spec_actor/effects.rs:117]: Addressed in `27e24ea0`: `counter_delta_from_params` → `Option<usize>`; Increment/DecrementByParam warn on missing/invalid params.

File: crates/temper-actor-runtime/src/spec_actor/effects.rs [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                        list.remove(idx);
+                    }
+                } else {
+                    tracing::warn!(
+                        actor = %self.name,
+                        list = %var,
+                        index_key = %index_key,
+                        "list_remove_at skipped: index param missing"
+                    );
+                }
```
[the remark this answers is not in the document]
Comment (nerdsane) [reply crates/temper-actor-runtime/src/spec_actor/effects.rs:178]: Addressed in `27e24ea0`: `counter_delta_from_params` → `Option<usize>`; Increment/DecrementByParam warn on missing/invalid params.

**Comment by nerdsane** [comment]
## ARENA SHIPPABLE · Grok · 2026-07-13 18:07 UTC

**PR:** https://github.com/nerdsane/temper/pull/380  
**HEAD:** `27e24ea07443` · branch `grok/arn-179-effect-vocabulary`  
**Linear:** ARN-179 · **ADR:** 0168  
**Merge:** nothing (arena rules)

### Checklist
| Gate | Evidence |
|------|----------|
| RED→GREEN TDD | list/counter regressions failed for stated reason; then 9/9 lib tests pass |
| Independent same-model review | **Verdict: PASS** — https://github.com/nerdsane/temper/pull/380#issuecomment-4960865457 |
| Greptile | Requested after PASS; P2 counter-delta warn fixed in `27e24ea0`; 4/4 threads resolved |
| Live E2E | https://github.com/nerdsane/temper/pull/380#issuecomment-4960876468 — `temper serve` → `/tdata/$metadata` 200; effect handle path tests green |
| CI green | https://github.com/nerdsane/temper/actions/runs/29271869000 — all required checks pass (Tests 17m47s) |

### Fix summary
- Pure-state effects implemented on PG actor-runtime path (`ListAppend`/`ListRemoveAt`/counter-from-param/by-param).
- Schedule/spawn fail closed at construction + `unreachable!` in apply.
- Serve gate allows pure-state; still rejects trigger/schedule/spawn.

**Comment by rita-aga** [comment]
@greptile review

**Comment by rita-aga** [comment]
@greptile review

**Comment by rita-aga** [comment]
Closing this ARN-179 competition entry. Adjudicated in favor of Fable #370 (completeness/scope). This branch is preserved for grafting its best parts into the winner; full scoring is on the ARN-165 arena board. (Rita-directed close, 2026-08-11.)
